experimental/displaynames/src/displaynames.rs

```diff
@@ -399,90 +453,184 @@ impl LocaleDisplayNamesFormatter {
+            .get_by(|uvstr| lang_id.strict_cmp(uvstr).reverse())
+    });
+
+    // At this point the key should exist in the language_data.
+    // Example: "en", "nl", "zh".
+    if ldn.is_none() {
+        ldn = match options.style {
+            Some(Style::Short) => language_data
                 .get()
                 .short_names
-                .get(&locale.id.language.into_tinystr().to_unvalidated()),
-            Some(Style::Long) => self
-                .language_data
+                .get(&lang_id.language.into_tinystr().to_unvalidated()),
+            Some(Style::Long) => language_data
                 .get()
                 .long_names
-                .get(&locale.id.language.into_tinystr().to_unvalidated()),
-            Some(Style::Menu) => self
-                .language_data
+                .get(&lang_id.language.into_tinystr().to_unvalidated()),
+            Some(Style::Menu) => language_data
                 .get()
                 .menu_names
-                .get(&locale.id.language.into_tinystr().to_unvalidated()),
+                .get(&lang_id.language.into_tinystr().to_unvalidated()),
             _ => None,
         }
         .or_else(|| {
-            self.language_data
+            language_data
                 .get()
                 .names
-                .get(&locale.id.language.into_tinystr().to_unvalidated())
+                .get(&lang_id.language.into_tinystr().to_unvalidated())
         });
+    }
+    // Throw an error if the LDN is none as it is not possible to have a locale string without the language.
+    return ldn.expect("cannot parse locale displayname.");
+}
 
-        if let Some(region) = locale.id.region {
-            let regiondisplay = match self.options.style {
-                Some(Style::Short) => self
-                    .region_data
+fn get_longest_qualifying_substrings<'a>(
+    locale: &Locale,
+    longest_matching_subtag: LongestMatchingSubtag,
+    locale_dn_formatter: &'a LocaleDisplayNamesFormatter,
+) -> Vec<&'a str> {
+    let LocaleDisplayNamesFormatter {
+        options,
+        region_data,
+        script_data,
+        variant_data,
+        ..
+    } = locale_dn_formatter;
+
+    let mut lqs: Vec<&str> = vec![];
+
+    if let Some(script) = locale.id.script {
+        // Ignore if the script was used to derive LDN.
+        if longest_matching_subtag != LongestMatchingSubtag::LangScript {
+            let scriptdisplay = match options.style {
+                Some(Style::Short) => script_data
+                    .get()
+                    .short_names
+                    .get(&script.into_tinystr().to_unvalidated()),
+                _ => None,
+            }
+            .or_else(|| {
+                script_data
+                    .get()
+                    .names
+                    .get(&script.into_tinystr().to_unvalidated())
+            });
+            if let Some(scriptdn) = scriptdisplay {
+                lqs.push(scriptdn);
+            }
```

Comment:
Use fallback, also for region below

```suggestion
            lqs.push(scriptdn.unwrap_or(script.as_str()));
```

Reply:
I started initially with adding the fallback, but couldn't get pass the borrow checker without using to_string().  
In this case,
```
{
   let mut lqs: Vec<&str> = vec![];
   if let Some(script) = locale.id.script {
       lqs.push(scriptdn.unwrap_or(script.as_str()));
   }
    return lqs;
}
```

Error:
"lqs.push(scriptdisplay.unwrap_or(script.as_str()));
                                                      --------------- `script` is borrowed here

return lqs;
    |                ^^^ returns a value referencing data owned by the current function"

Which makes sense. But I couldn't figure how to get around this without allocating a new string for script.

Reply:
pushed a fix